### ISSUE TITLE
Putting pressure on a vw daemon eventually results in model corruption

### ISSUE THREAD

**Issue by joachimdb** [JohnLangford/vowpal_wabbit#628]
This one is very hard to reproduce, but unfortunately also rather severe as it results in a corrupted model which cannot be recovered. (We've actually had this issue in production which we noticed because prediction performance suddenly dropped significantly.)

We made the following script for starting a vw daemon and doing some train/save cycles:

```
VW=/Users/joachim/opt/vowpal_wabbit-7.7/vowpalwabbit/vw
EXAMPLES_FILE=/Users/joachim/workspace/vw-socket-issue/resources/train-examples.txt
INITIAL_MODEL=/data/vowpalwabbit/sentiment.vw
FINAL_MODEL=/data/vowpalwabbit/sentiment-after.vw

## start vw
$VW -i $INITIAL_MODEL -f $FINAL_MODEL --save_resume --daemon --loss_function logistic --learning_rate 5.0 --decay_learning_rate 0.95 --power_t 0.5 --initial_t 0 --l1 1e-8 --l2 1e-8 

## do some train/save cycles, printing out the model size along the way:
## cycle 1:
$(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out1.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out2.txt &); $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out3.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out4.txt &)  ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out5.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out6.txt &) 

sleep 2
## save
echo "save" | nc -c localhost 26542
sleep 10

## print model size:
ls -l $FINAL_MODEL

sleep 5
## cycle 2
$(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out1.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out2.txt &); $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out3.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out4.txt &)  ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out5.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out6.txt &) 

sleep 1
## save
echo "save" | nc -c localhost 26542
sleep 10

## print model size:
ls -l $FINAL_MODEL

sleep 5

## cycle 3:

$(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out1.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out2.txt &); $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out3.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out4.txt &)  ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out5.txt &) ; $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out6.txt &) 

sleep 1
## save
echo "save" | nc -c localhost 26542
sleep 10

## print model size:
ls -l $FINAL_MODEL
```

Here's the shell output when running the script:

```
Joachims-MacBook-Air:vw-socket-issue joachim$ ./script.sh 
using l1 regularization = 1e-08
using l2 regularization = 1e-08
Num weight bits = 24
learning rate = 5
initial_t = 0
power_t = 0.5
final_regressor = /data/vowpalwabbit/sentiment-after.vw
using no cache
calling accept
calling accept
reading data from port 26542
calling accept
num sources = 1
calling accept
calling accept
calling accept
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
calling accept
calling accept
calling accept
calling accept
reading data from port 26542
write(net): Broken pipe
num sources = 1
reading data from port 26542
num sources = 1
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
saving regressor to /data/vowpalwabbit/sentiment-after.vw
calling accept
calling accept
calling accept
calling accept
calling accept
-rw-r--r--  1 joachim  wheel  43941545 May 12 15:05 /data/vowpalwabbit/sentiment-after.vw
reading data from port 26542
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
reading data from port 26542
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
saving regressor to /data/vowpalwabbit/sentiment-after.vw
0.333080   0.115549     11250470   2199516.5  -1.0000  -6.3057       15
0.333080   0.115549     11250470   2199516.5  -1.0000 -10.5392       21
calling accept
calling accept
calling accept
calling accept
calling accept
calling accept
-rw-r--r--  1 joachim  wheel  43941865 May 12 15:05 /data/vowpalwabbit/sentiment-after.vw
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
saving regressor to /data/vowpalwabbit/sentiment-after.vw
calling accept
calling accept
write error
calling accept
calling accept
calling accept
calling accept
-rw-r--r--  1 joachim  wheel  18203513 May 12 15:05 /data/vowpalwabbit/sentiment-after.vw
Joachims-MacBook-Air:vw-socket-issue joachim$ echo "save" | nc -c localhost 26542 && sleep 5 && ls -l /data/vowpalwabbit/sentiment-after.vw
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
saving regressor to /data/vowpalwabbit/sentiment-after.vw
-rw-r--r--  1 joachim  wheel  18205017 May 12 15:06 /data/vowpalwabbit/sentiment-after.vw
```

As can be seen, the model size is around 42MB after the first two save commands. The final save command results in a model size of 17MB, as do all subsequent save commands.

Notes:
* I haven't been able to reproduce this issue with a small initial model file. I can provide the model we used if wished.
* I used vw version 7.4 for the simple reason that the latest vw version 7.10.1 doesn't accept the older model
* There are some i/o errors that may or may not have to do with this particular issue (there is a `write(net): Broken pipe` error during the first cycle and a `write error` during the third cycle)
* The model size reduction always seems to occur after the third cycle only. I don't know why. 
* So far I haven't been able to reproduce this issue when l1 regularization is turned off
* see also https://groups.yahoo.com/neo/groups/vowpal_wabbit/conversations/topics/5064

**Comment by joachimdb** [comment]
Update: the issue really seems to occur only when the number of requested connections exceeds the number of vw children.

Thus, the same issue occurs with this script:

```
VW=/Users/joachim/opt/vowpal_wabbit-7.7/vowpalwabbit/vw
PID_FILE=/tmp/vw.pid
EXAMPLES_FILE=/Users/joachim/workspace/vw-socket-issue/resources/train-examples.txt
INITIAL_MODEL=/data/vowpalwabbit/sentiment.vw
FINAL_MODEL=/data/vowpalwabbit/sentiment-after.vw

## start vw
$VW --num_children 1 -i $INITIAL_MODEL -f $FINAL_MODEL --save_resume --daemon --loss_function logistic --learning_rate 5.0 --decay_learning_rate 0.95 --power_t 0.5 --l1 1e-8 --l2 1e-8 --initial_t 0 --pid_file $PID_FILE

## first cycle
for i in {1..2}
do
    $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out$i.txt &) 
done
sleep 3

## save and print model size:
echo "save" | nc -c localhost 26542
sleep 10
ls -lh $FINAL_MODEL

## second cycle:
for i in {1..2}
do
    $(nc -c localhost 26542 < $EXAMPLES_FILE > /tmp/out$i.txt &) 
done

## save and print model size:
echo "save" | nc -c localhost 26542
sleep 10
ls -lh $FINAL_MODEL
```

The resulting output now is:

```
Joachims-MacBook-Air:vw-socket-issue joachim$ ./script.sh 
using l1 regularization = 1e-08
using l2 regularization = 1e-08
Num weight bits = 24
learning rate = 5
initial_t = 0
power_t = 0.5
final_regressor = /data/vowpalwabbit/sentiment-after.vw
using no cache
calling accept
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
calling accept
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
calling accept
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
saving regressor to /data/vowpalwabbit/sentiment-after.vw
-rw-r--r--  1 joachim  wheel    42M May 12 16:56 /data/vowpalwabbit/sentiment-after.vw
calling accept
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
calling accept
reading data from port 26542
num sources = 1
average    since         example     example  current  current  current
loss       last          counter      weight    label  predict features
saving regressor to /data/vowpalwabbit/sentiment-after.vw
-rw-r--r--  1 joachim  wheel    27M May 12 16:56 /data/vowpalwabbit/sentiment-after.vw
```

**Comment by JohnLangford** [comment]
Using daemon mode with multiple simultaneous _training_ connections is 
inherently iffy, because they run without doing any locking on the 
shared model. As a consequence, all sorts of update rule confusion can 
occur.  It mostly works, but this should not be mistaken as safe because 
the order of operations cannot be assured.  Furthermore, I see no good 
way to debug when things go wrong with multiple simultaneous training 
connections because the result is not reproducible.

Multiple simultaneous testing connections should be fine and should 
produce reproducible results.  Having a single training connection and 
multiple testing connections should work ok, with the result of the 
training connection being reproducible.

I'd recommend against --decay_learning_rate in a daemon mode, because it 
changes the learning rate with the number of connections which seems 
plausibly undesirable.

-John

**Comment by mmroden** [comment]
How do you specify a single training connection and multiple testing
connections?

**Comment by JohnLangford** [comment]
Testing connections are connections which have no label information.

-John

**Comment by joachimdb** [comment]
Thanks for the update John. Just to be sure: by "connections which have no label info", do you mean examples without labels? What if importances are zero, is that safe?

I'm actually starting to wonder if there may be other ways in which we are abusing vw. For instance, I assumed that it is OK to send several prediction requests over the same connection (i.e. perform several writelines on the socket) before reading the responses. I assumed this is OK because one can match responses to requests via the example tags, but now I'm not sure anymore?

**Comment by JohnLangford** [comment]
On 05/26/2015 06:34 AM, Joachim De Beule wrote:
>
> Thanks for the update John. Just to be sure: by "connections which 
> have no label info", do you mean examples without labels? What if 
> importances are zero, is that safe?
>

Importance zero is not safe in general, because the --normalized part of 
the default update rule might adjust according to a new observed range 
of features.  For testing, you either want -t on the commandline or to 
not have a label.

> I'm actually starting to wonder if there may be other ways in which we 
> are abusing vw. For instance, I assumed that it is OK to send several 
> prediction requests over the same connection (i.e. perform several 
> writelines on the socket) before reading the responses. I assumed this 
> is OK because one can match responses to requests via the example 
> tags, but now I'm not sure anymore?
>

This should be fine, as long as it is not done to excess.  VW has an 
internal buffer of example (256? by default) so some asynchrony is 
required.  You don't even need tags, because VW always replies in order.

-John

**Comment by mmroden** [comment]
Would locking the model make sense here as well?  I've been thinking about
how to implement the locking for saving, but it seems that any
writable touching of the model file in daemon mode should be locked, given
the two bugs.

**Comment by JohnLangford** [comment]
Potentially, but it would need to be an option because some people like 
to use daemon mode for hog-wild style training, and that can be effective.

-John

**Comment by mmroden** [comment]
Gotcha.  Honestly sounds like 'hog wild' means 'could be acting entirely
unpredictably', but I get that this is currently default behavior.

Do you have any preferences about what the switch would look like?  I was
thinking '-ds/--daemon-safe' (instead of thread-safe).

**Comment by JohnLangford** [comment]
I'd think we should be safe by default and unsafe with an option.  Maybe 
"--no_lock".

-John

**Comment by joachimdb** [comment]
If I understand correctly the worst that can happen with hogwild! style training is basically that some updates are being overwritten etc., but not that a model is corrupted. If that is also the case with VW then I'm fine with that. Is it?

**Comment by JohnLangford** [comment]
It seems clear that things could go wrong when saving multiple models 
with regularization do to the way that implicit regularization works.
For training it works in practice, but there is no easy way to provide a 
guarantee.

-John

**Comment by pmineiro** [comment]
"So far I haven't been able to reproduce this issue when l1 regularization is turned off"

My guess is that the implementation of the strategy employed for l1 regularization in vw ("online gravity + rounding-at-the-end-of-training" http://www.jmlr.org/papers/volume10/langford09a/langford09a.pdf) is not consistent with the fast-and-loose hogwild strategy coupled with multiple writing connections and issuing intermediate saves.   

Lines like this (https://github.com/JohnLangford/vowpal_wabbit/blob/f62ea8df15ebfdcd994015daa7c813d426d23138/vowpalwabbit/gd.cc#L516) and this (https://github.com/JohnLangford/vowpal_wabbit/blob/f62ea8df15ebfdcd994015daa7c813d426d23138/vowpalwabbit/gd.cc#L552) give you some idea of what you are up against.

Either a detailed understanding of gd.cc would be required to introduce the requisite fine-grained locking, or a global lock could be introduced (but then, why use daemon mode?), or you could experiment with not saving until the end, or you could experiment with not using L1.

**Comment by pmineiro** [comment]
@joachimdb ... I suspect that the lazy projection for l1 in ftrl_proximal might behave better under concurrent updates than the default "online gravity+rounding" strategy.  Could you do some experiments using ftrl_proximal as the base learner and see if it improves things?

**Comment by JohnLangford** [comment]
Closing this for now.
